barretenberg/cpp/src/barretenberg/boomerang_value_detection/graph.cpp

```diff
@@ -0,0 +1,672 @@
+    if (!variables_vector.empty()) {
+        if (is_sorted_variables) {
+            for (size_t i = 0; i < variables_vector.size() - 1; i++) {
+                if (variables_vector[i] != ultra_circuit_builder.zero_idx &&
+                    variables_vector[i + 1] != ultra_circuit_builder.zero_idx &&
+                    variables_vector[i] != variables_vector[i + 1]) {
+                    {
+                        bool first_variable_is_not_constant =
+                            check_is_not_constant_variable(ultra_circuit_builder, variables_vector[i]);
+                        bool second_variable_is_not_constant =
+                            check_is_not_constant_variable(ultra_circuit_builder, variables_vector[i + 1]);
+                        if (first_variable_is_not_constant && second_variable_is_not_constant) {
+                            add_new_edge(variables_vector[i], variables_vector[i + 1]);
+                        }
+                    }
+                }
+            }
+        } else {
+            for (size_t i = 0; i < variables_vector.size() - 1; i++) {
+                for (size_t j = 1; j < variables_vector.size(); j++) {
+                    if (variables_vector[i] != ultra_circuit_builder.zero_idx &&
+                        variables_vector[j] != ultra_circuit_builder.zero_idx &&
+                        variables_vector[i] != variables_vector[j]) {
+
+                        bool first_variable_is_not_constant =
+                            check_is_not_constant_variable(ultra_circuit_builder, variables_vector[i]);
+                        bool second_variable_is_not_constant =
+                            check_is_not_constant_variable(ultra_circuit_builder, variables_vector[j]);
+                        if (first_variable_is_not_constant && second_variable_is_not_constant) {
+                            add_new_edge(variables_vector[i], variables_vector[j]);
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+template <typename FF>
+void Graph_<FF>::add_new_edge(const uint32_t& first_variable_index, const uint32_t& second_variable_index)
+{
+    variable_adjacency_lists[first_variable_index].emplace_back(second_variable_index);
+    variable_adjacency_lists[second_variable_index].emplace_back(first_variable_index);
+    variables_degree[first_variable_index] += 1;
+    variables_degree[second_variable_index] += 1;
+}
+
+template <typename FF> std::vector<uint32_t> Graph_<FF>::get_variable_adjacency_list(const uint32_t& variable_index)
+{
+    return variable_adjacency_lists[variable_index];
+}
+
+template <typename FF>
+void Graph_<FF>::depth_first_search(const uint32_t& variable_index,
+                                    std::unordered_set<uint32_t>& is_used,
+                                    std::vector<uint32_t>& connected_component)
+{
+    // this method realizes algorithm depth_first_search for undirected graph using the give variable
```

Comment:
`Implements` fits a little bit better here. Btw, whats the `give variable`?

Reply:
It's done